src/server/pps/server/internal_test.go

```diff
@@ -51,3 +51,196 @@ func TestNewMessageFilterFunc(t *testing.T) {
+                  "name": "7QGNJ3REq7OW6",
+                  "repo": "f3j93anf3bcyjkoidv12qlq7l8s8fpb0j8okz8x",
+                  "glob": "/fhgptj/(*)/(*)/(*)/(MU5YH7n7om78|LEHR3K01o9va|I77WHKzlepzg)",
+                  "joinOn": "$1/$2/$3"
+                }
+              },
+              {
+                "pfs": {
+                  "name": "LQIUnZQ24WFQvNJKB",
+                  "repo": "dzb49cse84hat6xbjqgiudxm5hw9gp",
+                  "glob": "/(*)/(*)/(*)",
+                  "joinOn": "$1/$2/$3"
+                }
+              }
+            ]
+          }
+        ]
+      }
+    ]
+  },
+  "autoscaling": true
+}`,
+			`{
+  "pipeline": {
+    "name": "iki30apxti994kpi1wnsko"
+  },
+  "transform": {
+    "image": "plmxrfvb9h4nsfjvc3nj56o3tu3kjtarxlzqpz4mb",
+    "cmd": [
+      "/bin/bash"
+    ],
+    "stdin": [
+      "#!/bin/bash",
+      "ffqu4wcxeh39iwa4s6mls2wx6uu6f3taappoxqs90ptdu57"
+    ],
+    "env": {
+      "LOG_LEVEL": "DEBUG",
+      "OUT_PATH": "/pfs/out",
+      "DJOLV20H89Z3INX6SF": "gxjewzd40kciifcjw4xi4lg97kuxji37rtme0e092o169sveco7ul",
+      "JMVVi64RAT0lKEQBr9GI08": "q",
+      "WTHVpL1VOo6JCBQ": "s",
+      "6GAX2AEZ8N54WPMQ": "u",
+      "8ZA60EP0yCCXEF": "y",
+      "3XKNeQ0G0GCGJk99RVE": "c",
+      "MJF2929EPwUP184": "5",
+      "C5M3WYLY9E5I6QBxSJENl463Q1": "i",
+      "3CW6X6WWjXJL2u2VWIH": "t",
+      "JAI0684D8XP1ELeSE43N": "v",
+      "W6MDSQEDkKZXw6EGUY": "r",
+      "PF0KP4PV5XAG2X": "9",
+      "A2NP6bP6M1hA4RQy00Y13": "f"
+    }
+  },
+  "parallelism_spec": {
+    "constant": "4"
+  },
+  "input": {
+    "cross": [
+      {
+        "pfs": {
+          "name": "H56IMlYJKJr8K5U",
+          "repo": "xte9lujxlfz",
+          "glob": "/ya9d0d"
+        }
+      },
+      {
+        "union": [
+          {
+            "pfs": {
+              "name": "ETO2oHU37",
+              "repo": "txz932n7gxao0szqmymep7vb56j44",
+              "glob": "/ggigb9/(*)/(*)/(*)",
+              "joinOn": "$1/$2/$3"
+            }
+          },
+          {
+            "join": [
+              {
+                "pfs": {
+                  "name": "3KWHBGKKl42D6",
+                  "repo": "kzfzxzkjv02pwimihxxl5y0ihwa6gvwqcpt5of9",
+                  "glob": "/9sgzyc/(*)/(*)/(*)/(TDH39N8n6f98|0YES2Yldixlj|IW4IWFlwof2r)",
+                  "joinOn": "$1/$2/$3"
+                }
+              },
+              {
+                "pfs": {
+                  "name": "LXKVmRYBHBF6pLLG9",
+                  "repo": "vems13rvb1zzj97xg57mef57dt9655",
+                  "glob": "/(*)/(*)/(*)",
+                  "joinOn": "$1/$2/$3"
+                }
+              }
+            ]
+          }
+        ]
+      }
+    ]
+  },
+  "autoscaling": true
+}`,
+		}
+	)
+	for i := 0; i < b.N; i++ {
+		// merge each spec with each other spec, and vice-versa; O(len(specs))
+		for _, defaults := range defaults {
+			for _, spec := range specs {
+				if _, _, err := makeEffectiveSpec(defaults, spec); err != nil {
```

Comment:
should this be wrapped in b.Run() so that we compare the bench for specific test-cases in the matrix?

Reply:
This is the current output:

```
$ go test -bench=. ./src/server/pps/server/ -run='^#'
goos: linux
goarch: amd64
pkg: github.com/pachyderm/pachyderm/v2/src/server/pps/server
cpu: 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
BenchmarkMakeEffectiveSpec-16                928           1284999 ns/op
PASS
ok      github.com/pachyderm/pachyderm/v2/src/server/pps/server 1.359s
```

Here it is with some more test cases, and split out by default and spec. What do you think?

```
$ go test -bench=. ./src/server/pps/server/ -run='^#'
goos: linux
goarch: amd64
pkg: github.com/pachyderm/pachyderm/v2/src/server/pps/server
cpu: 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
BenchmarkMakeEffectiveSpec/empty_default_createPipelineRequest/complex_spec_2-16                    9475            106834 ns/op
BenchmarkMakeEffectiveSpec/empty_default_createPipelineRequest/empty_spec-16                      234798              4828 ns/op
BenchmarkMakeEffectiveSpec/empty_default_createPipelineRequest/simple_spec-16                      40563             28351 ns/op
BenchmarkMakeEffectiveSpec/empty_default_createPipelineRequest/complex_spec-16                     10000            106277 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling/empty_spec-16                                      185751              6452 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling/simple_spec-16                                      41228             28368 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling/complex_spec-16                                     11286            105421 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling/complex_spec_2-16                                   11187            105802 ns/op
BenchmarkMakeEffectiveSpec/default_datum_tries/empty_spec-16                                      179964              6716 ns/op
BenchmarkMakeEffectiveSpec/default_datum_tries/simple_spec-16                                      42069             28752 ns/op
BenchmarkMakeEffectiveSpec/default_datum_tries/complex_spec-16                                     10000            105757 ns/op
BenchmarkMakeEffectiveSpec/default_datum_tries/complex_spec_2-16                                   10000            104657 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling_and_datumTries/complex_spec-16                       9619            105193 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling_and_datumTries/complex_spec_2-16                    10000            107963 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling_and_datumTries/empty_spec-16                       155136              7977 ns/op
BenchmarkMakeEffectiveSpec/default_autoscaling_and_datumTries/simple_spec-16                       40557             30176 ns/op
BenchmarkMakeEffectiveSpec/default_resource_requests_and_limits/empty_spec-16                      85954             13860 ns/op
BenchmarkMakeEffectiveSpec/default_resource_requests_and_limits/simple_spec-16                     35710             33154 ns/op
BenchmarkMakeEffectiveSpec/default_resource_requests_and_limits/complex_spec-16                    10000            111773 ns/op
BenchmarkMakeEffectiveSpec/default_resource_requests_and_limits/complex_spec_2-16                   9138            110678 ns/op
BenchmarkMakeEffectiveSpec/empty_defaults/complex_spec_2-16                                        11684            102862 ns/op
BenchmarkMakeEffectiveSpec/empty_defaults/empty_spec-16                                           238856              4535 ns/op
BenchmarkMakeEffectiveSpec/empty_defaults/simple_spec-16                                           44022             26786 ns/op
BenchmarkMakeEffectiveSpec/empty_defaults/complex_spec-16                                          10000            102784 ns/op
PASS
ok      github.com/pachyderm/pachyderm/v2/src/server/pps/server 33.064s
```